### PR TITLE
Wire diff orchestrator into editor validator (#321 step 5)

### PR DESCRIPTION
**Pull request by JamesPHoughton**
## Summary

Wires the full #321 validation pipeline into the editor's `validateTreatmentFile`. **First user-visible behavior change** in this series — earlier PRs added library helpers; this is what actually shows up in the squiggles.

## What the editor now does

`validateTreatmentFile` is async and runs this pipeline on each debounced edit:

1. **YAML parse + duplicate-key warnings** (existing behavior, preserved)
2. **Load imports** via `vscode.workspace.fs` (was sync before)
3. **Pre-hydration semantic checks** ([#324](https://github.com/deliberation-lab/stagebook/pull/324)) — template-name resolution + circular invocation detection
4. **Diff orchestrator** ([#327](https://github.com/deliberation-lab/stagebook/pull/327)) — runs schema twice, classifies issues as `matched` / `sourceOnly` / `hydratedOnly`
5. **Strict per-treatment reachable-keys check** ([#328](https://github.com/deliberation-lab/stagebook/pull/328)) — surfaces cross-treatment leaks and other unreachable-reference cases
6. **File-existence checks** (existing async behavior, preserved)

## Diagnostic display strategy

| Pipeline output | Editor display |
|---|---|
| `matched` (real bug in both passes) | red squiggle at source position |
| `unreachableReferences` (cross-treatment / uninvoked-template / etc.) | red squiggle at source position (walk-up if hydrated path doesn't resolve in source) |
| Pre-hydration semantic + hydration errors + YAML parse errors | red squiggle at source position |
| `sourceOnly` (mixed — artifact OR template-def bug not surviving expansion) | **yellow squiggle** at source position |
| `hydratedOnly` | not surfaced in source file; expanded preview already shows them |

The `sourceOnly` bucket gets warning-level treatment per the honest framing established in #327 — it genuinely can't be distinguished between artifacts and real bugs without provenance. Users see them but the validator doesn't overclaim.

## Path mapping

Source-pass issues (`matched`) land at their actual source paths via the existing AST mapper. `unreachableReferences` carry hydrated paths (e.g., paths inside a template-expanded treatment that don't exist in the unfilled source); we walk up the path one segment at a time until it resolves, landing at the closest source ancestor — typically the template invocation node for templated treatments. Last resort: top-of-file.

## Bundle fix

Also fixed a packaging bug in `runValidationDiff` — it imported `parse` from the `yaml` package, which uses `process` and isn't browser-safe. The VS Code webview broke with *"Dynamic require of 'process' is not supported"* (caught during smoke testing). Stagebook standardizes on `js-yaml` (browser-safe); switched to `js-yaml`'s `load`. Webview bundle returns to its previous size (728kb).

## Test plan

- [x] 7 new tests in `validateTreatmentDiff.test.ts` covering happy path, cross-treatment leaks, sourceOnly → warning routing, import-failure top-of-file errors, circular-template flagging, YAML parse error preservation
- [x] All 208 vscode + 87 viewer + 975 stagebook tests pass
- [x] Workspace lint + format clean
- [x] **Manual smoke test on pilot_3** ✓ — preview works; webview renders; no false-positive squiggles on the cleaned-up file
- [x] Extension reinstalled locally and verified

## Relationship to #321

Step 5 of the pipeline. Building on:
- [#324](https://github.com/deliberation-lab/stagebook/pull/324) — pre-hydration semantic (template-name resolution, circular invocations)
- [#327](https://github.com/deliberation-lab/stagebook/pull/327) — diff orchestrator + Zod-issue normalization
- [#328](https://github.com/deliberation-lab/stagebook/pull/328) — strict per-treatment unreachable-reference check
- [#329](https://github.com/deliberation-lab/stagebook/pull/329) — regression test for the intro-producer bug

Next: Phase 6 (static sweep against treatment files to catch any residual edge cases) and Phase 7 (docs + CHANGELOG + minor version bump).

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: apps/vscode/src/lib/validateTreatmentDiff.ts [hunk 1]
```diff
@@ -0,0 +1,229 @@
...
+  if (!loadResult.ok) {
+    // Couldn't even load imports. Surface as a top-of-file error;
+    // skip the rest of the pipeline (it'd just produce noise).
+    diagnostics.push({
+      message: loadResult.message,
+      severity: "error",
+      range: { startLine: 0, startCol: 0, endLine: 0, endCol: 1 },
+    });
+    return { diagnostics, parsedObj };
+  }
```
[the remark this answers is not in the document]
Comment (JamesPHoughton) [reply apps/vscode/src/lib/validateTreatmentDiff.ts]: Fixed in 462603b. Now short-circuits after `extractYamlErrors` if any error-level (non-duplicate-key) diagnostic was emitted — the import loader (and everything downstream) is skipped for a malformed root, so no redundant top-of-file "YAML parse error: …" duplicate. New test `short-circuits before import loading when YAML is malformed (no generic top-of-file dup)` covers it.

File: apps/vscode/src/lib/validateTreatmentDiff.ts [hunk 2]
```diff
@@ -0,0 +1,229 @@
...
+  if (diff.hydrationError) {
+    // Hydration failed. Top-of-file error; skip the diff buckets
+    // (they're empty anyway when hydration fails).
+    diagnostics.push({
+      message: diff.hydrationError,
+      severity: "error",
+      range: { startLine: 0, startCol: 0, endLine: 0, endCol: 1 },
+    });
+    return { diagnostics, parsedObj };
+  }
```
[the remark this answers is not in the document]
Comment (JamesPHoughton) [reply apps/vscode/src/lib/validateTreatmentDiff.ts:145]: Fixed in 462603b. When pre-hydration semantic already explained the root cause (e.g., "Template 'foo' is not defined"), the generic hydration error is suppressed — the specific diagnostic is the actionable one, and the generic top-of-file message would just be noise. New test `suppresses generic hydration error when pre-hydration semantic already explained it` verifies the dedup.

File: apps/vscode/src/extension.ts [hunk 3]
```diff
@@ -99,13 +97,29 @@ function toVscodeRange(range: Diagnostic["range"]): vscode.Range {
...
 
   const source = document.getText();
-  const result = validateTreatmentSource(source);
+  const rootDir = vscode.Uri.joinPath(document.uri, "..");
+  const decoder = new TextDecoder();
+
+  const result = await validateTreatmentWithDiff({
+    source,
+    loadImport: async (importPath: string) => {
+      const importUri = vscode.Uri.joinPath(rootDir, importPath);
```
[the remark this answers is not in the document]
Comment (JamesPHoughton) [reply apps/vscode/src/extension.ts]: Real concern, fixed in 462603b. Added an `isWithinWorkspace(importUri.fsPath, workspaceFolder.uri.fsPath)` guard in the `loadImport` callback. Out-of-workspace import paths throw a controlled error ("Import path escapes workspace: …") that surfaces as a diagnostic, instead of triggering arbitrary filesystem reads on every edit during validation. The try/catch wrapper around the validator body (also added in this commit, addressing your other comment) catches that error and routes it cleanly.

File: apps/vscode/src/extension.ts [hunk 4]
```diff
@@ -99,13 +97,29 @@ function toVscodeRange(range: Diagnostic["range"]): vscode.Range {
...
+      const bytes = await vscode.workspace.fs.readFile(importUri);
+      return decoder.decode(bytes);
+    },
+  });
```
[the remark this answers is not in the document]
Comment (JamesPHoughton) [reply apps/vscode/src/extension.ts]: Fixed in 462603b. Wrapped the `validateTreatmentWithDiff` call in try/catch. Unexpected failures surface as a top-of-file "Stagebook validator failed unexpectedly: …" diagnostic so the editor stays responsive instead of generating unhandled-promise warnings. Also returns early on stale-version mismatch within the catch path to avoid clobbering newer validation results.

File: apps/vscode/src/lib/validateTreatmentDiff.ts [hunk 5]
```diff
@@ -0,0 +1,229 @@
...
+  // survive expansion. Surface as warning so the user has visibility
+  // without us overclaiming "this is definitely a bug." The
+  // expanded-preview document still shows the full hydrated-pass
+  // diagnostics for users who want to drill in.
+  for (const issue of diff.sourceOnly) {
+    diagnostics.push({
+      message: appendPathIfMissing(issue),
+      severity: "warning",
+      range: resolveOrWalkUp(mapper, issue.path),
+    });
```
[the remark this answers is not in the document]
Comment (JamesPHoughton) [reply apps/vscode/src/lib/validateTreatmentDiff.ts:179]: Caught this one well. Fixed in 462603b — added `resolveIssueRange` that detects `code === "custom"` + `params.badKey` and uses `mapper.resolveKey` so the squiggle (and the quick-fix's `replace(diagnostic.range, suggestion)`) targets the key token. Falls through to value-range `resolveOrWalkUp` when `resolveKey` can't pin a range. Mirrors the existing `validateTreatmentSource` logic. New test `emits unrecognized-key issues with a key-token range (not value range)` verifies the column lands on the key character, not the value.
